Components: Add create-styles

sarayourfriend opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
The first step of #30503.

## How has this been tested?
There are minimal run-time changes from the `create-styles` in G2:
* Removal of `is` usage
* Localization of dependencies from `utils` into the package itself (basically getting rid of imports from `@wp-g2/utils`)
* Fixing some eslint errors (nested ternaries mostly)

Otherwise most of the changes are related to type issues that popped up for one reason or another and were fixed here.

## Types of changes
Adding the `create-styles` folder to `packages/components/src/ui`. This is a non-breaking change and, in fact, the new stuff isn't used anywhere yet so there's zero risk to this PR.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [N/A] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [N/A] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

sarayourfriend committed:
Components: Add create-styles

sarayourfriend committed:
Fix disallowed syntaxes

youknowriad commented on packages/components/package.json:
```diff
@@ -29,10 +29,13 @@
 		"@emotion/styled": "^10.0.23",
 		"@wordpress/a11y": "file:../a11y",
 		"@wordpress/compose": "file:../compose",
+		"@wordpress/data": "file:../data",
```
Ideally, the components package shouldn't rely on the data packages. What's the reason for this?

youknowriad commented on packages/components/package.json:
```diff
@@ -50,20 +53,26 @@
 		"moment": "^2.22.1",
 		"re-resizable": "^6.4.0",
 		"react-dates": "^17.1.1",
+		"react-merge-refs": "^1.1.0",
```
We have our own merge ref hook in the compose package that is better than this one because it handles dependencies better.

sarayourfriend committed:
Add missing TS resolution

youknowriad commented on packages/components/src/ui/create-styles/components/theme-provider/theme-provider-context.js:
```diff
@@ -0,0 +1,90 @@
+	/** @type {ThemeProviderContext} */ ( {
+		isDark: null,
+		isColorBlind: null,
+		isReducedMotion: null,
```
We have our own way of checking whether we're in reduced motion mode, I feel like we should remove this from there and rely on the existing hook instead (also in compose package). In other words, why is this a "theme" property?

youknowriad commented on packages/components/src/ui/create-styles/components/theme-provider/theme-provider-context.js:
```diff
@@ -0,0 +1,90 @@
+		isDark: null,
+		isColorBlind: null,
+		isReducedMotion: null,
+		isHighContrast: null,
```
Same question here, is this something we could retrieve from the browser (like we do for the  reduced motion hook) or is it something to be defined by the theme?

youknowriad commented on packages/components/src/ui/create-styles/components/theme-provider/theme-provider-context.js:
```diff
@@ -0,0 +1,90 @@
+ * @param {ThemeProviderContext} currentContextState
+ * @return {Record<string, string | boolean>} HTML attributes to apply for the current theme provider context state.
+ */
+export function useThemeProviderModeHtmlAttributes( currentContextState = {} ) {
```
I don't understand why we need both this `useThemeProviderModeHtmlAttributes` and `useThemeProviderContextBridge` hooks? They sound like something that components that consume the theme should do and not something the framework provides (reading more to try to understand)?

youknowriad commented on packages/components/src/ui/create-styles/components/theme-provider/theme-provider-context.js:
```diff
@@ -0,0 +1,90 @@
+	} )
+);
+
+export const useThemeProviderContext = () => useContext( ThemeProviderContext );
```
could this be just `useTheme` maybe?

youknowriad commented on packages/components/src/ui/create-styles/components/theme-provider/theme-provider.js:
```diff
@@ -0,0 +1,170 @@
+	);
+
+	return (
+		<div
```
Why do we need a "div" for the theme provider?

youknowriad replied:
In other words, in my mind, the theme provider just sets a "context" and let the components consume the data from that theme. so I'd like to understand more why we need a div. We tend to be as pure as possible and try to kill divs as much as we can in the editor and Gutenberg related screens.

youknowriad commented on packages/components/src/ui/create-styles/components/theme-provider/theme-provider.js:
```diff
@@ -0,0 +1,170 @@
+	useColorBlindMode( { isColorBlind, isGlobal, ref: nodeRef } );
+	useDarkMode( { isDark, isGlobal, ref: nodeRef } );
+	useHighContrastMode( { isGlobal, isHighContrast, ref: nodeRef } );
+	useReducedMotionMode( { isGlobal, isReducedMotion, ref: nodeRef } );
```
These all look like hooks that set "data attributes" on the div? Why is that?

youknowriad reviewed (commented):
Thanks for starting this PR. I didn't do a full review but I have some questions. 

Overall, I'm trying to see if we can iterate and simplify things. I'm basically trying to understand how the theme provider works and questioning why we need all these things :P. At first sight it looks way more complex than what I have imagined.

sarayourfriend committed:
Remove ThemeProvider and duplicated hooks

sarayourfriend commented:
Thanks @youknowriad. After our discussion I've deleted the ThemeProvider and the duplicated hooks 🎉

youknowriad commented on packages/components/package.json:
```diff
@@ -50,12 +52,16 @@
+		"create-emotion": "^10.0.27",
 		"dom-scroll-into-view": "^1.2.1",
 		"downshift": "^6.0.15",
+		"emotion-theming": "^10.0.27",
```
This seems useless now right?

youknowriad commented on packages/components/src/ui/create-styles/create-compiler/create-compiler.js:
```diff
@@ -0,0 +1,121 @@
+/**
+ * External dependencies
+ */
+import { is } from '@wp-g2/utils';
```
This import/dependency probably something we can avoid? (use lodash helpers or just regular JS checks)

sarayourfriend replied:
Whoops, yes, looks like I missed one 😅

youknowriad commented on packages/components/src/ui/create-styles/create-compiler/create-compiler.js:
```diff
@@ -0,0 +1,121 @@
+		/* eslint-enable jsdoc/valid-types */
+		{
+			__insert.apply( customEmotionInstance.sheet, [ ...args ] );
+			customEmotionInstance.__events.emit( 'sheet.insert', ...args );
```
I didn't see this event consumed anywhere? do we still need this? for what?

sarayourfriend replied:
This is part of the custom iFrame support described here: 

https://github.com/ItsJonQ/g2/blob/97ef23ebe56a4968701a36ce6697f75803700134/packages/styles/src/components/README.md#L20

It is consumed by the `StyleFrameProvider` component here: 

https://github.com/ItsJonQ/g2/blob/97ef23ebe56a4968701a36ce6697f75803700134/packages/styles/src/components/style-frame-provider.js#L199

youknowriad commented on packages/components/src/ui/create-styles/create-compiler/plugins/index.js:
```diff
@@ -0,0 +1,37 @@
+ * Internal dependencies
+ */
+import cssVariablesPlugin from './css-variables';
+import specificityPlugin from './extra-specificity';
```
Can we add a comment to clarify that we can drop the plugins once IE11 is no longer supported?

sarayourfriend replied:
Yup! That's actually already called out in the README but I can move it to the top of the document to make it clearer.

youknowriad commented on packages/components/src/ui/create-styles/create-style-system/README.md:
```diff
@@ -0,0 +1,110 @@
+
+Component polymorphism is documented via the `PolymorphicComponent` type in `polymorphic-component.ts`.
+
+The same module also exposes a `ViewOwnProps` type which gives the `css` and `as` props to a regular `Props` interface:
```
Should we avoid the `css` prop?

youknowriad commented on packages/components/src/ui/create-styles/create-style-system/constants.js:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Uses the the prefix for the CSS variables compiled by the system.
+ */
+export const NAMESPACE = '--wp-g2';
```
this should probably be `--wp-admin` like our existing CSS variables but maybe we can keep this temporary prefix until we remove the css variables.

sarayourfriend replied:
Yes, let's put that off until we've culled the number of variables, it'll make it easier to make the distinction when debugging.

youknowriad commented on packages/components/src/ui/create-styles/create-style-system/constants.js:
```diff
@@ -0,0 +1,22 @@
+
+export const MODE_SPECIFICITY_COMPOUND_LEVEL = 3;
+
+export const INTERPOLATION_CLASS_NAME = '__interpolationClassName__';
```
Seems like some of the variables here are not useful anymore now that we dropped the ThemeProvider right?

sarayourfriend replied:
I think you're right! I'll make sure I remove the unused ones.

sarayourfriend replied:
Actually never mind, these are all used by the global style system which is the basis of the theme variables for the new style system (as in, the foundation for how those variables get transformed from an object passed to `createStyleSystem` and then turned into actually applied CSS custom properties)

youknowriad replied:
Some of the variables here refer explicitly to the "data" attributes that were used previously in the theme provider, do we still need these?

youknowriad commented on packages/components/src/ui/create-styles/create-style-system/create-core-element.js:
```diff
@@ -0,0 +1,182 @@
+ * createCoreElement is a super light-weight higher-order wrapper that
+ * enhances base elements (like `div`, `input`, or even Components) with
+ * features provided by the Style system, such as direct CSS compiling via
+ * the `css` prop.
```
Can we remove this?

sarayourfriend replied:
Just remove the documentation or support for the `css` prop? The `css` prop is used by the component system internally (and it's a powerful tool that `emotion` exposes itself as well). Even if we decide not to have it be a documented external API I think we should leave documentation about why `createCoreElement` exists, right? Otherwise it will be a mystery even to those contributing to `@wordpress/components` what this `css` prop is or where it comes from for `styled` components.

youknowriad replied:
I thought (and read somewhere in the docs here) that internally the styled components work by just supporting the "className" prop, isn't that true? In which cases, we do need the css prop? I'm asking cause if we allow it in some components but not others, it might be harder to distinguish between "internal" or "public" components.

sarayourfriend replied:
So to be entirely honest up until now I hadn't done an audit of the `css` prop's usage, but it's actually super minimal outside of stories and tests in G2.

I believe we could remove it altogether with minimal disruption (we'll have to re-write some stories as they come in but that's not a big deal).

The `css` prop is merely a convenience prop that allows you to pass styles inline, similar to the `style` prop but more powerful because it leverages all of the style system in there. However, you can basically do the same thing with the `className` prop so I'm not sure I'm convinced the `css` prop is actually a necessity (it might even be an anti-pattern like inline styles tend to be).

youknowriad commented on packages/components/src/ui/create-styles/create-style-system/create-core-element.js:
```diff
@@ -0,0 +1,182 @@
+ * the `css` prop.
+ *
+ * A styled element also has built-in baseStyles (which can be adjusted using
+ * the createStyleSystem factory).
```
How is this useful?

sarayourfriend replied:
For example, here:

https://github.com/WordPress/gutenberg/blob/ea4544d971cb1e355721e192236719e364a06ae3/packages/components/src/ui/create-styles/create-style-system/create-core-element.js#L72-L77

I think that's a good usage of the base-styles, ensuring that all style system components automagically comply with reduced motion.

youknowriad replied:
I do agree these styles are good defaults (both border box and reduced motion) but I'm not sure it's worth the complexity specially if it's the only thing added by this function.

sarayourfriend replied:
This function does more than just add the base styles:

* Adds base styles
* Adds `css` function to `styled` components (used by the component system)
* Adds the `interpolationClassname` to `styled` components
* Automatically handles prop filtering so that invalid props passed to `styled` components don't automatically get forwarded even if they're used by the `styled` component in a props function (although we don't use this in practice so I think we should remove it, it's also risky to me anyway)

sarayourfriend replied:
After my most recent push that leaves us with: 

* Adds `css` function to `styled` components (used by the component system)
* Adds the `interpolationClassname` to `styled` components

The `interpolationClassname` is such a powerful and simple feature, we should _not_ sacrifice it. It actually is a pretty good performance benefit to have it as the only alternative solution to it is to use React context which will just add more nodes to the React tree OR to hand-write classnames into CSS-in-JS but that is _not_ a common pattern and should be recommended against strongly unless there's no other way to do things.

The base-styles I'm unattached to, but if we're already doing this for the interpolation classname we might as well leave the base styles too.

youknowriad commented on packages/components/src/ui/create-styles/create-style-system/create-core-element.js:
```diff
@@ -0,0 +1,182 @@
+ * @param {CreateCoreElementOptions} options Options to custom coreElement styles.
+ * @return {import('./polymorphic-component').PolymorphicComponent<TTagName, {}>} The Style system wrapped HTMLElement/React.Component.
+ */
+export const createCoreElement = ( tagName, options ) => {
```
I'm actually wondering whether this is really necessary at all

sarayourfriend replied:
We can talk about it here: https://github.com/WordPress/gutenberg/pull/30509#discussion_r609709604

youknowriad commented on packages/components/src/ui/create-styles/create-style-system/create-styled-components.js:
```diff
@@ -0,0 +1,175 @@
+ * context connection at all. This is HUGE for performance as there are far less
+ * React.Component nodes within the render tree.
+ *
+ * This is thanks to how the Style system compiles and coordinates style values.
```
This sound like these are "shortcuts" to using raw `styled` right? Are we using these? do we need them?

sarayourfriend replied:
I'm not following your question. What do you mean by a "shortcut" and "raw `styled`"?

youknowriad replied:
Nevermind :)

youknowriad commented on packages/components/src/ui/create-styles/create-style-system/generate-theme.js:
```diff
@@ -0,0 +1,84 @@
+ * @param {GenerateThemeProps} props Props to generate a Style system theme with.
+ * @return {GenerateThemeResults} A set of variables and content for the System.
+ */
+export function generateTheme( {
```
Is this still useful now that we removed the theme provider?

sarayourfriend replied:
Yes, for globally applied styles.

youknowriad replied:
> for globally applied styles.

What are these?

sarayourfriend replied:
This discussion: https://github.com/WordPress/gutenberg/pull/30509#discussion_r607829644 is about the same thing.

youknowriad commented on packages/components/src/ui/create-styles/create-style-system/utils.js:
```diff
@@ -0,0 +1,186 @@
+ * @param {boolean} isGlobal
+ * @return {string} Compiled innerHTML styles to be injected into a <style /> tag.
+ */
+export function transformValuesToVariablesString(
```
This sound like things used by the ThemeProvider to generate the CSS variables, since we don't have that anymore, maybe there are some things to clean/remove here?

sarayourfriend replied:
Nope, this is used by `generateTheme` to apply the generated theme variables to the `:root` scope at the very least as well as the different "modes" offered by the new style system.

youknowriad replied:
so basically, this is just a duplication of what we currently do in this mixin https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_mixins.scss#L441 

I think we should just add the G2 variables there for now and avoid all the extra JS logic that runs at init.

sarayourfriend replied:
But aren't we trying to do CSS-in-JS? It seems counter intuitive that we would keep our variables in SASS in that case?

youknowriad replied:
There's a few things here:

- I think the logic of these function is just too complex for what it is right now, why not just inject a static `<style>` element? 
- I'd like us to avoid having two ways of injecting these "theme styles", so we should consolidate in one way
- Right now the variables that constitute a "theme" are used not just in the "@wordpress/components" package, they can be used in all the other packages and these packages are not CSS in JS.

youknowriad replied:
In other words, If we really want to keep it in JS (which I don't feel strongly against), I think it should be:

 - its own package that all the other packages that use the variables should depend on. (Right now the dependency towards the SASS mixin is in the build script)
 - Simplify it's logic to just print the style properly.
 - Remove the SASS mixin. 

I feel it's a lot of work probably, more than just moving the G2 variables to the mixin for the moment.

sarayourfriend replied:
I think the last one is the most compelling reason to me.

I think it'd still be good to have the `ui.get` function to make variables easy to retrieve (and have the variables documented in JS and strictly typed so that people aren't accidentally introducing new variables or using non-existent ones)...

I'll have to think about how to do this but I agree we can remove this variable generation stuff 🤔 I just wish the original author were around to give their expert opinion on this 😞

sarayourfriend replied:
> Simplify it's logic to just print the style properly.

By this do you mean that the new package would just create the theme variables into the `:root` scope?

I'm not sure which is less work. Refactoring how the theme is created is an unknown variable to me.

youknowriad replied:
yes, basically do what `_default-custom-properties.scss` does now.

sarayourfriend replied:
Where would the new package's function to create the theme variables be called?

youknowriad replied:
I think my ideal solution for now, is just to remove all the theme generation, grab the CSS variables injected right now and move them to the existing mixing and call it a day. (assuming we work on a follow-up on removing the need for most of these variables and replace them with static values like discussed).

youknowriad replied:
I'd love other thoughts on this as well. This is something that can have some impact long term (WP user profiles, theming components) so yeah more thoughts and opinions here would be good. cc @mtias @jasmussen @mcsf @gziolo

sarayourfriend replied:
The thing is, there are a TON of variables that are generated. I'm not sure it's possible to just plop them into a mixin as some of them are runtime generated using `tinycolor`: 
https://github.com/ItsJonQ/g2/blob/97ef23ebe56a4968701a36ce6697f75803700134/packages/styles/src/theme/utils.js#L45

https://github.com/ItsJonQ/g2/blob/97ef23ebe56a4968701a36ce6697f75803700134/packages/styles/src/theme/tokens/colors.js#L196

https://github.com/ItsJonQ/g2/blob/97ef23ebe56a4968701a36ce6697f75803700134/packages/styles/src/theme/tokens/colors.js#L83

youknowriad replied:
@sarayourfriend would it make sense in this case to actually switch the priorities of the integration:

 1- First refactor the `ui.get` to avoid returning variables (and just return static values potentially from a theme context) instead of returning CSS variables (the main color can be kept as a variable)
 2- Do this removal later.


How much work would **1** here be. It feels to me that's important regardless because it prevents us from introducing these public variables inadvertently.

sarayourfriend replied:
So `ui.get` actually does just return the CSS custom property. Get is a one line function here: 

https://github.com/WordPress/gutenberg/blob/da56a34c0697e8f7dae3139ee7952e597f9e63de/packages/components/src/ui/create-styles/create-style-system/create-style-system.js#L138-L142

```ts
function get(
  key: keyof TConfig | keyof TDarkConfig | keyof THCConfig | keyof TDarkHCConfig | TGeneratedTokens
) {
  return `var(${ createToken( key.toString() ) })`;
}
```

As you can see it just calls `createToken` on the passed in `key` and generates the `var(...)`.

So is what you're proposing to change `get` to actually return the raw value of the variable rather than the CSS Custom Property? Just want to make sure I'm understanding.

I think this could be relatively simple... the theme is a flat object so I think `get` would just be turned into this:

```ts
function get(
  key: keyof TConfig | keyof TDarkConfig | keyof THCConfig | keyof TDarkHCConfig | TGeneratedTokens
) {
  return themeConfig[key];
}
```

Is that right?

youknowriad replied:
yes something like that, though I do wonder where this `themeConfig` come from :P

youknowriad replied:
In my mind, the "themeConfig" here is context aware (so back to the ThemeProvider just setting a context value)

sarayourfriend replied:
🤔 I'll need to think about this.

My initial thought is just that it would pluck it from the passed in `config` but that leaves out the different "modes". Not sure what to do about that as the previous way modes were handled was to just swap out the variables and the CSS would update automatically, but if we're returning raw values from `get` that's not going to be possible.

Likewise, `get` needs to be usable outside the context of React (like in a `styles.js` file where we're just calling `css` and `styled`) so we cannot rely on all those values being inside of a React context (plus adding React context everywhere adds a performance hit that I'm pretty sure Q spent a good deal of time measuring).

mcsf replied:
> I'd love other thoughts on this as well. This is something that can have some impact long term (WP user profiles, theming components) so yeah more thoughts and opinions here would be good.

I lean towards playing nicely with existing systems and starting small. As Riad points out, the impact of these changes is unclear, and there's the risk of starting _de facto_ APIs that will later cost us to support.

sarayourfriend replied:
What if we change to prefix to `--wp-experimental` for all these variables, leave the rest the way it is, and then re-evaluate how to do this once we've integrated the `styles` package which actually contains the theme? It might be more helpful to brainstorm solutions at that point than now when we only have half the context.

jasmussen replied:
I would defer to Riad or Miguel, whose concerns I share: we should move carefully and only introduce components, systems, and custom properties when we know for a fact that they will be needed and used. 

To use an example: color custom properties. I've seen a tendency to design big palettes and adding them all as custom properties to be conveniently available. In practice, they end up being dead weight on most pages, and causing developer confusion with the breadth of choice (maybe gray-200 was chosen when gray-300 should've been used).

That is to say: these properties are only ever useful if they will be used, and ideally used often enough to deserve such stature. I do recognize that it's a spectrum ranging from overwhelming palettes to lack of flexibility, and that the ideal spot, probably, is somewhere in the middle. But it is an argument for moving slowly and only when we have some confidence in the system.

gziolo replied:
[@wordpress/base-styles](https://github.com/WordPress/gutenberg/tree/trunk/packages/base-styles) contain several files and a huge amount of code. Ensuring it stays in sync with CSS-in-JS bits might be a huge challenge unless we find a good way to share them. 

> By this do you mean that the new package would just create the theme variables into the :root scope?

This sounds like a reasonable proposal because we could use those variables in other places – outside of `@wordpress/components`.

youknowriad replied:
> What if we change to prefix to --wp-experimental for all these variables, leave the rest the way it is, and then re-evaluate how to do this once we've integrated the styles package which actually contains the theme? It might be more helpful to brainstorm solutions at that point than now when we only have half the context.

That's fine I guess but it's definitely something we need to address.

youknowriad commented on packages/components/src/ui/create-styles/hooks/use-hydrate-global-styles.js:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @param {UseHydrateGlobalStylesProps} props Props for the hook.
+ */
+export function useHydrateGlobalStyles( { globalStyles, injectGlobal } ) {
```
This also sounds like something that is only useful for the theme provider and that can be removed.

sarayourfriend replied:
Same as https://github.com/WordPress/gutenberg/pull/30509#discussion_r607829644

gziolo commented on packages/components/src/ui/create-styles/create-compiler/README.md:
```diff
@@ -0,0 +1,68 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+-   [create-compiler](#create-compiler)
+    -   [Breakpoint values](#breakpoint-values)
+    -   [Plugins](#plugins)
+    -   [Custom iframe support](#custom-iframe-support)
+    -   [Interplated Components](#interplated-components)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
```
We probably can remove this section completely here and in other places:
```suggestion
```

youknowriad reviewed (commented):
I saw a few G2 mentions in the docs code..... I think it's something we should remove at some point as it's just a code name, but we can do so in a follow-up if needed.

sarayourfriend committed:
Clean up READMEs

sarayourfriend committed:
Remove unnecessary deps

sarayourfriend committed:
Remove stray `is` import

sarayourfriend committed:
Change namespace to experimental to avoid introducing new public APIs

sarayourfriend committed:
Remove stray mentions of g2

sarayourfriend committed:
Remove `css` prop and automatic prop filtering

sarayourfriend commented:
After the discussions above I've:

1. Removed the `css` prop from `styled` components
2. Removed automatic prop filtering from `styled` components
3. Changed the CSS custom properties prefix to `--wp-experimental` so we're clear about the status of these APIs

youknowriad reviewed (approved):
👍 Let's move to the next step
